### ISSUE TITLE
Zed Python doesn't detect multiple venvs in project

### ISSUE THREAD

**Issue by KhazAkar** [zed-industries/zed#20402]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I have monorepository with few sub-projects, each have their own venv. Currently it only sees venv in the root of monorepo, but doesn't detect venvs in subfolders.

### Environment

Zed: v0.160.7 (Zed)
OS: Linux Wayland ubuntu 22.04
Memory: 15.2 GiB
Architecture: x86_64
GPU: Intel(R) UHD Graphics 620 (WHL GT2) || Intel open-source Mesa driver || Mesa 23.2.1-1ubuntu3.1~22.04.2

**Comment by osiewicz** [comment]
Yup, that's a current limitation; each project has multiple worktrees and each worktree has at most one associated venv. You can work around it by opening your monorepo with multiple worktrees.

I realize it's not ideal, but we're aware of that.

**Comment by KhazAkar** [comment]
Okay, so this issue can stay as tracking of eventual progress towards fixing that limitation. As of now, I will try opening subdirs :)

**Comment by osiewicz** [comment]
Right; one problem we've had with allowing arbitrary venvs in the project is that currently we only allow using a single language server per worktree (so same as single venv per worktree), so having multiple venvs in a single project seemed like a lot of work of unclear benefit. I'd love to learn about your use case more; does the limitation of *language servers* having one instance per worktree bother you anyhow?

**Comment by KhazAkar** [comment]
Well, currently I work in monorepository, which contains few projects, quite coupled together, and they use Python, JS and C++. Some of those projects use python, but each with different version and libraries needed, so either we use poetry, or venv, depending on a project. More than 1 of them uses venv, most notably main directory (for tools and CI), and one of the projects inside one of the subfolders.
In this case, technically it bothers me, because I need, currently as limitation, to either open multiple zed instances, or use micromamba for environments and swap between them, depending on what part of monorepo I work currently on.

**Comment by andres-ortizl** [comment]
Hello ! 

I currently work in a monorepo as well, which contains several projects that are fairly independent but share some common code. These projects primarily use Python but are configured with different virtual environments (venv) or dependency managers (e.g., Poetry). This setup often requires switching between environments based on the file or project I'm working on.

The limitation I’ve found, specifically in Zed (and also in VS Code to some extent), is the inability to dynamically switch the active virtual environment based on the file or folder that’s currently focused. For example, in Python, I have to manually adjust the pyright configuration or switch environments when working on different parts of the monorepo. This is tedious when jumping between projects as part of my normal workflow.

In VS Code, this limitation is addressed with an extension like [Python Envy](https://marketplace.visualstudio.com/items?itemName=teticio.python-envy), which dynamically adjusts the virtual environment based on the workspace or file being edited. I think adding a similar feature to Zed would be a game-changer for managing Python projects in a monorepo setup.

**Comment by osiewicz** [comment]
This is somewhat related to https://github.com/zed-industries/zed/pull/22182 as we'd have to understand at which point in your worktree a given subproject starts.
Would either of you say that each subdirectory that you'd like to use distinct virtual environment for can be clearly "recognized" as a Python project? e.g. do these subdirectories have pyproject.toml files in them? Or do they meet any other criteria?

**Comment by KhazAkar** [comment]
@osiewicz most of the time it's designated by having pyproject.toml file. One exception I know of are python azure functions, but for now we can stick with it

**Comment by andres-ortizl** [comment]
I found a partial workaround for dynamically updating the pyrightconfig.json file with the virtual environment path corresponding to the current project when switching between projects. Here’s how I solved it:

I created a small Python script that, combined with a Zed task, automates updating the pyrightconfig.json file based on the virtual environment (.venv) for the current file’s project (if a .venv exists).

How It Works : 
1.	Zed Task Configuration:
I added the following task to my Zed configuration, which calls the Python script and passes the directory of the current file ($ZED_DIRNAME) as an argument:

```json
 {
    "label": "UPDATE pyright",
    "command": "python",
    "args": [
      "path_to_script.py",
      "$ZED_DIRNAME"
    ],
    "use_new_terminal": false,
    "allow_concurrent_runs": false,
    "reveal": "always",
    "hide": "never",
    "shell": "system"
  }
```

2.	Python Script:
The Python script performs the following steps:
	•	Reverse Traversal: It starts from the given directory (passed via $ZED_DIRNAME) and traverses upwards until it finds a pyproject.toml file, which identifies the root of the project.
	•	Locate .venv: If a .venv folder exists in the project directory, it updates the pyrightconfig.json file to point to the .venv path.
	•	Update Configuration: The script either modifies an existing pyrightconfig.json file or creates one if it doesn’t already exist, adding the correct venvPath and venv fields.

```python

import json
import os


def find_pyproject_toml(start_path: str):
    current_dir = start_path
    while True:
        potential_path = os.path.join(current_dir, "pyproject.toml")
        print(f"Checking {potential_path}")
        if os.path.isfile(potential_path):
            return current_dir
        parent_dir = os.path.dirname(current_dir)
        if parent_dir == current_dir:
            return None
        current_dir = parent_dir


def update_pyright_config(base_dir, pyright_config_path):
    venv_path = os.path.join(
        base_dir, ".venv", "bin", "python"
    )

    if not os.path.isfile(venv_path):
        print(f"Python interpreter not found at {venv_path}")
        return

    pyright_config = {}

    if os.path.isfile(pyright_config_path):
        with open(pyright_config_path, "r") as file:
            pyright_config = json.load(file)
    pyright_config["venvPath"] = base_dir
    pyright_config["venv"] = ".venv"

    with open(pyright_config_path, "w") as file:
        json.dump(pyright_config, file, indent=4)

    print(f"Pyright configuration updated with virtual environment from {venv_path}")


if __name__ == "__main__":
    import sys

    print(sys.argv[1:])
    start_path = os.getenv("ZED_FILE")
    print(f"Searching for pyproject.toml starting from {start_path}")
    base_dir = find_pyproject_toml(start_path)
    print(f"Found base directory at {base_dir}")
    update_pyright_config(
        base_dir=base_dir,
        pyright_config_path=os.path.join(
            "/root_project_path/", "pyrightconfig.json" # change me
        ),
    )
```

**Comment by orokusaki** [comment]
Another monorepo user here. I've just started a new job and we're using a monorepo with five different `pyproject.toml` / `.venv` in separate app / service subdirectories.

I'm primarily adding this as a +1 / bump, but also so I'm not notified on further comments.

**Comment by MarcoGlauser** [comment]
Also ran into this when trying to switch from PyCharm to Zed. 
Spent hours trying to figure out why pyright wouldn't pick the execution environments defined in the pyrightconfig.
Pyright does support multiple execution environments in one source tree:
https://microsoft.github.io/pyright/#/configuration?id=execution-environment-options

I'll go back to PyCharm in the meantime, until Zed is more mature.

**Comment by michael-ud** [comment]
@osiewicz any thoughts on whether this will still end up rolling out soon? Once it does I'll have a stronger case to convice peeps on my team to give Zed a try, because they're accustumed to constantly venv-switching in their editors.

**Comment by osiewicz** [comment]
We've landed some changes recently; give it a crack and let me know how it goes. We're gonna work on improving our Python story in the near-term future.

**Comment by matthew-robb** [comment]
> We've landed some changes recently; give it a crack and let me know how it goes. We're gonna work on improving our Python story in the near-term future.

Thanks for the heads up. I can now select the appropriate venv using the search feature, however, it is not resolving any of my imports for installed dependencies. My settings.json is empty.

**Comment by orokusaki** [comment]
@osiewicz I can second what @matthew-robb said. If I open Zed in the top level directory in our project, and then I open a python module in a sub directory a few layers down, sometimes the .venv above that module will activate automatically, while other times my global Python version will be used even though there is a uv.lock and pyproject.toml and a .venv. But in all cases imports are not resolved, and attempts to import from package packages install installed in the .venv don't have completions.

**Comment by SeanMooney** [comment]
i tried to use zed in a proejct that uses tox to execute ptyhon tests.
event if i type out the full path to the tox venvs i.e. `.tox/py3/bin/python` in the venv selector zed seams to ignore it.

that works fine in things like vscode so if feel liek just allowing users to pass the path to the ptyhon binary in the vnev woudl allow people to specyify how they wanted this to work.

auto discovery is nice but just being able to say "use this" is also helpful especially fi i want to debug things in different python verions.

at least in the flatpack version venv supprot seams to still be a limitation of zed ux
Zed 0.196.5 
cc82f1eacd880a8730db24c448e922898601c849

**Comment by SeanMooney** [comment]
by the way i have both a .venv that  i use to install tox pre-commit and git-reivew or othr tools not part of my project

and also have the vnevs created by tox in a .tox folder at the root of my directory

so the venvs i want to use are named after the tox envoinmoent that we use for local testing and ci.

the auto detection works at the top level and find my .venv that i created for managing tools but zed seam to have a alot of artificial limitations on which path it can use that make using it with opensouce proejct where we cant expect venv to be at the root impossible.

i could work around this by just manulaly installing the deps the same way tox would but most user wont actully put in the effort to figure out how to fix this type of issue if they just want to try a new editor to see if it will work for them in there workflow.

**Comment by osjerick** [comment]
I just noticed this today. I'm also working on a monorepo; in Zed, the venv is selected correctly, but it seems the language server doesn't use it. Something I do with PyCharm, which could be an idea to consider, is to open one of the subprojects and then **attach** (which is an option) the others I want to work with; the IDE treats them independently, the root of each is clear, while showing them together in the project panel.

**Comment by osiewicz** [comment]
Another ping; we allow selecting a venv per pyproject.toml detected within your project and as of 0.201.0 it should be respected by language servers (we now spawn multiple language servers in such cases).

**Comment by michael-ud** [comment]
@osiewicz thanks for the update.

Our repo looks like this:

```
src
    sub
        one
            pyproject.toml
            uv.lock
            .venv/
        two
            pyproject.toml
            uv.lock
            .venv/
            app/
                main.py
        three
            pyproject.toml
            uv.lock
            .venv/
            app/
                main.py
```
Ideally, we'd be able to do this:

```
cd ~/dev/alpha/src
zed .
```

And then when opening e.g., `sub/one/app/main.py` that tab would automatically use the Python in `sub/one/.venv`.

However, I downloaded the latest Zed update (0.201.4 64ec0f6a04f27d4e2e0ada9b9432df9c5340c84f) and tried it out in our monorepo by opening Zed at the root of the repo. When I opened a module within a directory that had a `pyproject.toml`/`uv.lock`/`.venv/` Zed remained on my global Python. I had to do what we've previously been doing, which is to `source sub/one/.venv/bin/activate` and then `zed .` to get it to pick up the venv. IOW, there doesn't appear to be any noticeable change.

**Comment by osiewicz** [comment]
@michael-ud thanks for taking the time to write up such a detailed response. It could be that Zed has persisted your choice of a global venv from "good old times" where we sucked at this use case. When you open a toolchain selector, the venvs should be listed correctly (or at least they are for me in the following repro):
```
#!/usr/bin/env bash
set -euo pipefail

BASE_DIR="src/sub"

# Directories to create
PROJECTS=("one" "two" "three")

for project in "${PROJECTS[@]}"; do
    PROJ_DIR="$BASE_DIR/$project"
    mkdir -p "$PROJ_DIR"

    # Add app dir + main.py only for two and three
    if [[ "$project" == "two" || "$project" == "three" ]]; then
        mkdir -p "$PROJ_DIR/app"
        cat > "$PROJ_DIR/app/main.py" <<EOF
def main():
    print("Hello from $project")

if __name__ == "__main__":
    main()
EOF
    fi

    # Minimal pyproject.toml
    cat > "$PROJ_DIR/pyproject.toml" <<EOF
[project]
name = "$project"
version = "0.1.0"
requires-python = ">=3.11"
dependencies = []
EOF

    # Create venv with uv
    (cd "$PROJ_DIR" && uv venv)

    # Generate lock file (uv will resolve empty deps fine)
    (cd "$PROJ_DIR" && uv lock)
done
```

https://github.com/user-attachments/assets/51fb132f-e379-4cd7-8173-ffe3ca705e4e

Once you select a virtual environment, it should remain selected even across restarts and so on. We do not "overrule" the selected virtual environment anyhow, even if there is a virtual environment within the root path of the current subproject. This is why you might be seeing a "global" venv take precedence.

**Comment by michael-ud** [comment]
@osjerick thanks for the response. I ran `rm -r ${HOME}/Library/Application\ Support/Zed/db` before trying it out, so it should have been a fresh start WRT state.

I actually failed to mention that we're using a UV workspace structure - we only have a `.venv/` in the top level (adjacent to a `pyproject.toml` and `uv.lock`:

```
/src
    /some-stuff
    /other-stuff
    /my-monorepo
    ├── pyproject.toml  # Defines the workspace and common dependencies
    ├── uv.lock         # Single lockfile for the entire workspace
    ├── .venv/...
    └── sub/
        ├── my-app/
        │   └── pyproject.toml  # Defines dependencies for `my-app`
       ├── .venv/...
        └── my-lib/
            └── pyproject.toml  # Defines dependencies for `my-lib`
```

So, what I noticed is that, if I cd into `src/my-monorepo` (the directory with the `uv.lock` and `.venv`) and open Zed from there, it works, but if I cd into `src` and open Zed from there, then it doesn't work (results in global venv selection and the .venv not showing up in the dropdown even when I open a module in the nested directories - screenshot below).

<img width="542" height="147" alt="Image" src="https://github.com/user-attachments/assets/f7088162-4fb9-48ec-ad0e-683e9e4a1168" />

**Comment by osiewicz** [comment]
Ok, sweet - I can repro this. Also, I know where the problem is - we just need to feed extra search paths to `pet` beyond the root of the workspace.

Btw. I assume that thr second `.venv` in your example is a copy-paste error? The one immediately above `my-lib`, that is.

**Comment by michael-ud** [comment]
@osiewicz oh good deal (re: being able to repro) - and yeah, that second `.venv` was a copy/paste error.

**Comment by osiewicz** [comment]
@michael-ud is your issue fixed? If so, I'll close this one optimistically.

**Comment by michael-ud** [comment]
@osiewicz sorry to keep you waiting; between our travel and some work deadlines, I never had the chance to manually install the nightly build, etc., but I just checked in Zed 0.202.7 (cd9b0b5be63f2ddde915e5e1c382318ddb4f2d1f) and the problem is not solved (opening in our `src` directory without manually activating a `venv` first results in the global Python interpreter). Is that merge not yet in production? I'll have the opportunity to do a manual install of a nightly build on Thursday, if not, and I can try it then.

**Comment by osiewicz** [comment]
It is in Preview currently AFAIK. It should be enough to install a Preview version (0.203.x)

**Comment by michael-ud** [comment]
@osiewicz I'm excited to report that this is working perfectly, as far as I can tell, as of `0.203.4` - thank you!

**Comment by osiewicz** [comment]
I'm glad to hear that! Let me know y'all if you run into any issues.

**Comment by GabrielBianconi** [comment]
LFG!!! thank you, works great

**Comment by osjerick** [comment]
Working great! Thank you!!
